File: fsharp/AdventOfCode.2021/AdventOfCode.2021/Challenges/Challenge03.fs
module Challenges.Challenge03

open Utils.IO
open System

let getMostCommon (arr:string array) i =
    arr
    |> Array.map (fun s -> Seq.item i s)
    |> Array.groupBy (fun s -> s)
    |> Array.sortWith
        (fun a b ->
            if Seq.length (snd a) = Seq.length (snd b) then
                int (fst b) - int (fst a)
            else
                (Seq.length (snd b) ) - (Seq.length (snd a))
        )
    |> Array.head
    |> fst

let findCode (data:string array) cmp =
    [| 0 .. 12  |]
    |> Array.fold
        (fun acc i ->
            match Array.length acc with
            | 1 -> acc
            | _ ->
                let mc = getMostCommon acc i
                Array.filter (fun s -> cmp (Seq.item i s) mc) acc) data
    |> Array.head

let setup =
    readLines<string> 3

let part1 (input:string array) =
    let len = Seq.length input.[0]
    let counts = [|
        for i in 0..len-1 do
            input
            |> Array.groupBy (fun x -> Seq.item i x)
            |> Array.sortBy fst
            |> Array.map (fun (k,v) -> Seq.length v) |]

    let gamma =
        seq {0..len-1}
            |> Seq.fold
                (fun a i ->
                    let count = Seq.item i counts
                    if count.[1] > count.[0] then
                        a + (1 <<< (len-1-i))
                    else
                        a
                ) 0

    let epsilon = ~~~gamma &&& 0xFFF
    gamma*epsilon |> string

let part2 input =
    let oxygen = findCode input (fun a b -> a = b)
    let co2 = findCode input (fun a b -> a <> b)
    Convert.ToInt32(string oxygen, 2) * Convert.ToInt32(string co2, 2) |> string

File: csharp/AdventOfCode.2021/Challenges/Challenge03.cs
using AdventOfCode.Core;
using AdventOfCode.Core.IO;

namespace AdventOfCode2021.Challenges;

[Challenge(3)]
public class Challenge03(IInputReader InputReader)
{
    [Part1]
    public async Task<string> Part1Async()
    {
        var data = await InputReader.ReadLinesAsync(3).ToArrayAsync();
        var len = data[0].Length;
        var counts = Enumerable.Range(0, len)
            .Select(i => data
                .Select(x => x[i])
                .GroupBy(x => x)
                .OrderBy(x => x.Key)
                .Select(x => x.Count())
                .ToArray())
            .ToList();

        var gamma = 0;

        for (var i = 0; i < counts.Count; i++)
            if (counts[i][1] > counts[i][0])
                gamma += 1 << (len - 1 - i);

        var epsilon = ~gamma & 0xFFF;

        return (gamma * epsilon).ToString();
    }

    [Part2]
    public async Task<string> Part2Async()
    {
        var data = await InputReader.ReadLinesAsync(3).ToArrayAsync();
        var oxygenList = data.ToList();
        var co2List = data.ToList();

        var len = data[0].Length;
        for (var i = 0; i < len; i++)
        {
            if (oxygenList.Count == 1 && co2List.Count == 1)
                break;

            var commonOxy = GetMostCommon(oxygenList, i);
            var commonCo2 = GetMostCommon(co2List, i);

            if (oxygenList.Count > 1)
                oxygenList = oxygenList.Where(x => x[i] == commonOxy).ToList();

            if (co2List.Count > 1)
                co2List = co2List.Where(x => x[i] != commonCo2).ToList();
        }

        var o = Convert.ToInt32(oxygenList[0], 2);
        var c = Convert.ToInt32(co2List[0], 2);

        return (o * c).ToString();
    }

    private static char GetMostCommon(IList<string> list, int position) => list
        .Select(x => x[position])
        .GroupBy(x => x)
        .OrderByDescending(x => x.Count())
        .ThenByDescending(x => x.Key)
        .Select(x => x.Key)
        .First();
}